### ISSUE TITLE
Umbrella issue: migrate kubernetes to Go 1.12.1

### ISSUE THREAD

**Issue by spiffxp** [kubernetes/kubernetes#75372]
The Go release team hopes to have 1.12.1 out sometime between today and tomorrow https://github.com/kubernetes/kubernetes/issues/74890#issuecomment-472558713

We believe this will fix an issue we're working around in one area of our upgrade tests, but may be lurking in all of our vendored code https://github.com/kubernetes/kubernetes/pull/75305

Let's use this issue to put together the steps to roll forward.  Here's what I gathered from #sig-testing, comment what I've left out or gotten wrong and I'll edit the issue description

>when 1.12.1   hits the road, looks like there are 3 pieces that we need to update?
> - rules_go (similar to https://github.com/bazelbuild/rules_go/pull/1963) (we will likely need help from @ixdy or @jayconrod here to actually get a PR merged, though we could open one)
> - kubekins (similar to https://github.com/kubernetes/test-infra/pull/11583)
> - kubecross (similar to https://github.com/kubernetes/kubernetes/pull/74632)

> also the kubernetes/kubernetes WORKSPACE file: https://github.com/kubernetes/kubernetes/blob/897d62ace79a8fa72f9825e065b2b67fdba65bec/build/root/WORKSPACE#L33 (but maybe you fold that into update rules_go)
> similar to https://github.com/kubernetes/kubernetes/pull/74632

/milestone v1.14
/priority critical-urgent
/kind cleanup

**Comment by spiffxp** [comment]
FYI @bradfitz
FYI @justinsb @liggitt @neolit123 @dims @tpepper 
I don't want to voluntold one of you to lead this but you've all got more context than I
/assign @amwat 
I know release team's test-infra role needs to be involved

/sig release
/sig testing

**Comment by spiffxp** [comment]
/assign @munnerz 
since @dims pinged you in #sig-testing

**Comment by ixdy** [comment]
@jayconrod is pretty good about publishing new `rules_go` releases as soon as a new upstream go version is released, but it's pretty much up to us to update our workspace dependencies.

**Comment by jayconrod** [comment]
bazelbuild/rules_go#1994 will add support for the new Go releases. I'll cherry-pick that back to the 0.18, 0.17, and 0.16 branches when CI is happy and tag new releases after that. Should be ready tomorrow.

**Comment by bradfitz** [comment]
Go 1.12.1 is out: https://golang.org/dl/

(And a Go 1.11.x update.)

**Comment by jayconrod** [comment]
rules_go [0.18.1](https://github.com/bazelbuild/rules_go/releases/tag/0.18.1), [0.17.2](https://github.com/bazelbuild/rules_go/releases/tag/0.17.2), and [0.16.8](https://github.com/bazelbuild/rules_go/releases/tag/0.16.8) are now out with support for the new Go versions.

**Comment by liggitt** [comment]
opened https://github.com/kubernetes/kubernetes/pull/75390 with the same set of changes made in https://github.com/kubernetes/kubernetes/pull/74632

build-cross needs to be built/pushed

**Comment by neolit123** [comment]
revert for the workaround is on hold here:
https://github.com/kubernetes/kubernetes/pull/75393

**Comment by liggitt** [comment]
looks like the golang:1.12.1 docker image isn't available yet. once that's available, we'll need to rebuild our cross-build image, then https://github.com/kubernetes/kubernetes/pull/75390 can proceed

**Comment by liggitt** [comment]
golang:1.12.1 image publish is pending at https://github.com/docker-library/official-images/pull/5550

**Comment by mariantalla** [comment]
List of related issues that we believe will be unblocked:
* https://github.com/kubernetes/kubernetes/issues/74890
* https://github.com/kubernetes/kubernetes/issues/75039
* https://github.com/kubernetes/kubernetes/issues/74887

**Comment by dims** [comment]
Plan of record after the sig-release burndown meeting:
- The 1.12.1 golang image is not yet ready (we are waiting on https://github.com/docker-library/official-images/pull/5550)
- We can't wait till Monday for the 1.12.1 image
- So, let's try hack up kube-cross build to download the 1.12.1 binaries while leaving `FROM golang:1.12.0` as-is
- @BenTheElder has the pen the hack and @dims to review
- @amwat will help publish the new kube-cross image and update test-infra
- @hoegaarden will check if there is any go references elsewhere in release tooling
- @spiffxp to write up the steps if we have to do the worst case scenario - "switch everything back to 1.11.5"
- NOTE: we will not go to 1.11.6 as that is a new unknown entity as well.

**Comment by hoegaarden** [comment]
> @hoegaarden will check if there is any go references elsewhere in release tooling

I couldn't find any reference to a specific version of go in anago/gcb/gcbmgr.

**Comment by BenTheElder** [comment]
Upgrade to go1.12.1 is almost ready to go, just waiting on the kube-cross image to get promoted https://github.com/kubernetes/kubernetes/pull/75413

**Comment by dims** [comment]
Here's the test-infra change from @amwat https://github.com/kubernetes/test-infra/pull/11804

**Comment by dims** [comment]
docker-library/official-images#5550 has merged, but @amwat @BenTheElder and i are going ahead with what we had planned earlier. 

@BenTheElder is in the middle of a CL request for kube-cross with a hacked Dockerfile in https://github.com/kubernetes/kubernetes/pull/75413

@amwat has a kubekins-test image, but that's under control (different GCS bucket)

**Comment by dims** [comment]
https://github.com/kubernetes/test-infra/pull/11804 has merged ... 

still waiting for kube-cross to reach gcr repo - https://console.cloud.google.com/gcr/images/google-containers/GLOBAL/kube-cross?gcrImageListsize=30

**Comment by dims** [comment]
we have the image now - https://console.cloud.google.com/gcr/images/google-containers/GLOBAL/kube-cross@sha256:f89b510d690188ffbbff18e1268e88d33642d8c95de8b589f6320085a954ad0c/details?tab=info

**Comment by dims** [comment]
`docker pull gcr.io/google-containers/kube-cross:v1.12.1-1` works ..
`docker pull k8s.gcr.io/kube-cross:v1.12.1-1` does not work yet ... probably need to wait for some more time 

`docker pull k8s.gcr.io/kube-cross:v1.12.0-1` works for sure (that's the existing image that we use)

**Comment by dims** [comment]
looks like the image made it all the way now

**Comment by dims** [comment]
unholding https://github.com/kubernetes/kubernetes/pull/75393 so that can merge

**Comment by dims** [comment]
waiting on pull-kubernetes-kubemark-e2e-gce-big to go green in https://github.com/kubernetes/kubernetes/pull/75413

**Comment by dims** [comment]
@liggitt can you please approve https://github.com/kubernetes/kubernetes/pull/75413 ?

**Comment by dims** [comment]
please see discussion at https://kubernetes.slack.com/archives/C2C40FMNF/p1552690316531000 to approve #75413

**Comment by neolit123** [comment]
@spiffxp @dims 

the SIGABRT is back after using 1.12.1:
```
I0316 02:37:10.603] fatal error: sync: inconsistent mutex state
I0316 02:37:10.606] 
I0316 02:37:10.606] goroutine 292 [running]:
I0316 02:37:10.606] runtime.throw(0x4b8393e, 0x1e)
I0316 02:37:10.606] 	/usr/local/go/src/runtime/panic.go:617 +0x72 fp=0xc002065e30 
...
```

https://prow.k8s.io/log?id=2847&job=ci-kubernetes-e2e-gce-new-master-upgrade-cluster
https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-new-master-upgrade-cluster

see my comment from here:
https://github.com/kubernetes/kubernetes/pull/75305#issuecomment-472426749

> so this could simply be a case of resource exhaustion for other unknown reasons, somehow matching the 1.12 switch timing.

for now, my suggesting is to apply our fix:
https://github.com/kubernetes/kubernetes/pull/75305

roll the release with 1.12.1 and investigate better fixes post-release.

**Comment by justinsb** [comment]
I'd hope we haven't had enough time yet @neolit123 but I'm trying to figure it out....

**Comment by tpepper** [comment]
We've finally got #75413 (updates to 1.12.1) and #75393 (revert workaround) merged.

The revert merged first and was caught in a CI run, which in a way nicely failed:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-cluster/2847
"fatal error: inconsistent mutex state"
This is nice because it is expected and again gives us a feeling the problem is common somehow in this one test case.

With the updates to 1.12.1 bits merged we're awaiting soak of this e2e over the weekend and watching for failures in test runs >= 2848.  We're specifically hoping for green runs and no "fatal error: inconsistent mutex state".

**Comment by neolit123** [comment]
ok, looks like https://prow.k8s.io/log?id=2848&job=ci-kubernetes-e2e-gce-new-master-upgrade-cluster
doesn't have the problem with go 1.12.1.

**Comment by dims** [comment]
Whew!!

**Comment by dims** [comment]
filed follow ups for monday
- https://github.com/kubernetes/test-infra/pull/11810
- https://github.com/kubernetes/kubernetes/pull/75422/

cc @spiffxp @BenTheElder @amwat @tpepper @neolit123 @ixdy

**Comment by spiffxp** [comment]
Checking in now that https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-cluster/2848 has completed.  The "inconsistent mutex state" failure didn't crop up. One flake that appears unrelated (has happened prior to go1.12 migration)

Agree we are clear to move to go1.12.1 the "correct" way.

**Comment by liggitt** [comment]
we're now on 1.12.1 official images

/close
